src/server/frontend_wayland/window_wl_surface_role.cpp

```diff
@@ -528,6 +528,9 @@ void mf::WindowWlSurfaceRole::handle_enter_output(graphics::DisplayConfiguration
                     client,
                     [&](OutputInstance* instance)
                     {
+                        if(auto fractional_scale = surface.value().get_fractional_scale())
+                            fractional_scale->output_entered(instance);
+
```

Comment:
I think this is the wrong spot for the `entered` call? We don't want to call `output_entered` once per global bound by the client, we just want to call it once, with the output that the surface has entered.

That would be up at line 526 - within the `if (config.id == id) { ... }` conditional, not within the `for_each_output_bound_by` functor.

This would probably mean changing the signature of `output_{entered,left}` to take a `DisplayConfigurationOutput const&`, and the `FractionalScaleV1` object would then want to have a set of `{id, scale}` instead of `OutputInstance*`.

Reply:
You are right. I incorrectly understood your explanation and thought that `handle_enter_output` was called once and `for_each_output_bound_by` would handle looping over the outputs that interest us (In retrospect, I have no idea why I thought of this). I just tried your suggestion and the code still works. Thanks!